Python bindings: Use ogr.DataSource and gdal.Dataset as context managers

dbaston opened the pull request:
## What does this PR do?

Allows the OGRDataSource proxy to serve as a context manager, so instead of:

```
ds = drv.CreateDataSource(fname)

lyr = ds.CreateLayer('test', geom_type = ogr.wkbPoint)

f = ogr.Feature(lyr.GetLayerDefn())
f.SetGeometry(ogr.CreateGeometryFromWkt("POINT (0 0)"))
lyr.CreateFeature(f)

del ds  # or ds = None, or ds.Destroy(), or ds.FlushCache()
```
(one of the "gotchas" documented at https://gdal.org/api/python_gotchas.html#saving-and-closing-datasets-datasources)

You can do 
```
with drv.CreateDataSource(fname) as ds:
    lyr = ds.CreateLayer('test', geom_type = ogr.wkbPoint)

    f = ogr.Feature(lyr.GetLayerDefn())
    f.SetGeometry(ogr.CreateGeometryFromWkt("POINT (0 0)"))
    lyr.CreateFeature(f)
```

I'm guessing there is a reason this hasn't been done...in which case this PR can document that reason :)

## What are related issues/pull requests?

## Tasklist

 - [x] Add test case(s)
 - [x] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

rouault commented:
> I'm guessing there is a reason this hasn't been done...

I can't really think of a reason why it hasn't been done.

There would still be a gotcha in scenarios like:
```
def create_layer():
     with drv.CreateDataSource("foo.shp") as ds:
            return ds.CreateLayer("foo")
```

where the return Python layer would point to a destroyed C++ OGRLayer, but this is not worse than current situation, so doesn't necessarily need to be addressed  (to be solved it would probably require the Python datasource to have a set of weak references to layer it has returned, and invalidates those Python layer objects)

dbaston commented:
One bad thing is you can get a crash by

```
with drv.CreateDataSource(fname) as ds:
    pass

lyr = ds.GetLayer(0)
```

To avoid this I thought about having `__exit__` call `FlushToCache` instead of `Destroy`. But then it looked like maybe that only works for certain drivers?

rouault commented:
> ```
> 
> ```
> 
> 
>     
>   
> 
> To avoid this I thought about having `__exit__` call `FlushToCache` instead of `Destroy`. But then it looked like maybe that only works for certain drivers?

FlushCache() doesn't close files, connections, etc.   Only GDALClose(), which is what Destroy() ultimately calls, does.
One possibility would be after calling Destroy() to also set the SWIG .thisown and .this member to None. Cf what ReleaseResultSet() does in 26abcb035b7bb854322c5b8ab7c89532f43d1c9b

sgillies commented:
Yes, I think making the layer's reference to the dataset a weak reference is the key. The Fiona project takes a different approach in that it only surfaces layers as collections, and skips over this issue.

What is the motivation for making the OGR bindings better? They've been terrible in this way for > 15 years and no one cared except me 😆

dbaston commented:
> One possibility would be after calling Destroy() to also set the SWIG .thisown and .this member to None

This is the magic I was looking for! Now the example above would give following exception:

```
Traceback (most recent call last):
  File "/tmp/test.py", line 22, in <module>
    lyr2 = ds.GetLayer(0)
  File "/home/dan/dev/build-gdal-Desktop-Debug/swig/python/osgeo/ogr.py", line 1232, in GetLayer
    return self.GetLayerByIndex(iLayer)
  File "/home/dan/dev/build-gdal-Desktop-Debug/swig/python/osgeo/ogr.py", line 967, in GetLayerByIndex
    return _ogr.DataSource_GetLayerByIndex(self, *args)
TypeError: in method 'DataSource_GetLayerByIndex', argument 1 of type 'OGRDataSourceShadow *'
```

Which is not an intuitive error, but way better than crashing.

> What is the motivation for making the OGR bindings better now? 

At the moment, just looking for small changes that can help make the tests more readable.

dbaston commented:
I've updated this to provide the same treatment for both `gdal.Dataset` and `ogr.DataSource`. It seems that setting `self.this = None` in the `__exit__` method is enough to destruct the underlying object, so I've removed the call to `self.Destroy()` and did not set `self.thisown = None`.

sgillies commented:
@dbaston using `self.Destroy()` would make the bindings more predictable. It would insure that the dataset is finalized and deallocated before any other code is executed. The only thing that `self.this = None` does is decrement the reference count of the object that was named `self.this`. Other code could be executed before Python's garbage collector realizes that the object can be deleted.

dbaston commented:
Thanks for pointing that out, @sgillies . I reverted to call `self.Destroy()` for `ogr.DataSource`. I don't see a corresponding method in `gdal.Dataset` so I called `_gdal.delete_Dataset(self)`.

SWIG is new to me so if it seems like I'm flailing it's because I am :)

rouault commented on autotest/gcore/basic_test.py:
```diff
@@ -756,3 +756,23 @@ def test_basic_test_DontUseExceptions():
+
+
+def test_create_context_manager(tmp_path):
+    numpy = pytest.importorskip("numpy")
```
```suggestion
```

rouault commented on autotest/gcore/basic_test.py:
```diff
@@ -756,3 +756,23 @@ def test_basic_test_DontUseExceptions():
+        ds.GetRasterBand(1).ReadAsArray()
+
+    ds_in = gdal.Open(fname)
+    dat_in = ds_in.GetRasterBand(1).ReadAsArray()
```
yes I try to avoid using the numpy dependency as much as possible (admitedly this is a bit painful at times not using it)

you could actually use Fill(100) instead of Fill(0), and check the GetRasterBand(1).Checksum() value  (which should be != 0 with a fill value != 0)

dbaston replied:
Got it, thanks

rouault commented:
Merging.
https://gdal.org/tutorials/raster_api_tut.html#using-create and https://gdal.org/tutorials/vector_api_tut.html#writing-to-ogr could potentially be updated to reflect the possible use of the context managers
